lib/services/workflow-api-service.js

```diff
@@ -143,11 +143,13 @@ function workflowApiServiceFactory(
             return taskGraphProtocol.cancelTaskGraph(instanceId);
         })
         .then(function(cancelledGraphId) {
-            if (cancelledGraphId !== instanceId) {
-                // Cancelled graph ID doesn't match the requested ID.
-                throw new Errors.TaskCancellationError('Failed to cancel ' + graphId);
-            }
-            return waterline.graphobjects.needByIdentifier(graphId);
+            return waterline.graphobjects.needByIdentifier(graphId)
+            .then (function(workflow) {
+                if (workflow.active()) {
+                    throw new Errors.TaskCancellationError('Failed to cancel ' + graphId);
+                }
+                return waterline.graphobjects.needByIdentifier(graphId);
```

Comment:
Graph cancellation is not instantaneous, it is an asynchronous action that gets added to the task scheduler queue. This seems likely to produce false negatives.